internal/rest/leveldbreceipts.go

```diff
@@ -0,0 +1,157 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rest
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/kaleido-io/ethconnect/internal/errors"
+	"github.com/kaleido-io/ethconnect/internal/kvstore"
+)
+
+type levelDBReceipts struct {
+	conf  *LevelDBReceiptStoreConf
+	store kvstore.KVStore
+}
+
+func newLevelDBReceipts(conf *LevelDBReceiptStoreConf) (*levelDBReceipts, error) {
+	store, err := kvstore.NewLDBKeyValueStore(conf.Path)
+	if err != nil {
+		return nil, errors.Errorf(errors.ReceiptStoreLevelDBConnect, err)
+	}
+	return &levelDBReceipts{
+		conf:  conf,
+		store: store,
+	}, nil
+}
+
+// AddReceipt processes an individual reply message, and contains all errors
+// To account for any transitory failures writing to mongoDB, it retries adding receipt with a backoff
+func (l *levelDBReceipts) AddReceipt(requestID string, receipt *map[string]interface{}) (err error) {
+	// insert an entry with a composite key to track the insertion order
+	timeElapsed := ((*receipt)["headers"].(map[string]interface{}))["timeElapsed"].(float64)
+	timeSeconds := (*receipt)["receivedAt"].(int64) + int64(timeElapsed)
+	timeStringSeconds := fmt.Sprintf("%d", timeSeconds)
+	timeStringNanoSeconds := fmt.Sprintf("%.9f", timeElapsed)
+	timeStringNanoSeconds = timeStringNanoSeconds[strings.Index(timeStringNanoSeconds, "."):]
+	timestamp := timeStringSeconds + timeStringNanoSeconds
+
+	// add "z" prefix so these entries come after the actual entries
+	// because for iteration we start from last backwards
+	compositeKey := fmt.Sprintf("z%s:%s", timestamp, requestID)
+	l.store.Put(compositeKey, nil)
+
+	// inser the actual entry
+	b, _ := json.MarshalIndent(receipt, "", "  ")
+	return l.store.Put(requestID, b)
+}
+
+// GetReceipts Returns recent receipts with skip & limit
+func (l *levelDBReceipts) GetReceipts(skip, limit int, ids []string, sinceEpochMS int64, from, to string) (*[]map[string]interface{}, error) {
+	itr := l.store.NewIterator()
+	defer itr.Release()
+	results := make([]map[string]interface{}, 0, limit)
+	index := 0
+	var errResult error
+	for valid := itr.Last(); valid; valid = itr.Prev() {
+		if limit > 0 && len(results) >= limit {
+			break
+		}
+
+		if index >= skip {
+			val := itr.Value()
+			if len(val) > 0 {
+				// the index keys don't have values. if we have hit an entry with a value
+				// then we have reached the end of the index keys
+				break
+			}
+			// extract the request key from the index key
+			r, _ := regexp.Compile("z[0-9.]+:(.+)")
```

Comment:
This is really performance sensitive code I think - potentially running millions of times for a single query, where we have a request that misses all records.
We should probably use a `split` if we can safely.
If we can't, we should definitely at least pre-compile the regexp.

Reply:
... came back to this as I was making another comment. I think we are overcomplicating things, and again it will have a performance implication. Please see notes on line 57 above